internal/os/os.go

```diff
@@ -1,12 +1,260 @@
+		o.RequestedOsCommit = newOSInfo.CommitID
+		cmd := exec.Command("rpm-ostree", "update", "--preview")
+		stdout, err := cmd.Output()
+
+		if err != nil {
+			log.Error("Failed to run 'rpm-ostree update --preview'")
+			return err
+		}
+
+		// TODO check - if there is a multiple commits?
+		if !strings.Contains(string(stdout), newOSInfo.CommitID){
+			log.Error("Failed to run 'rpm-ostree update --preview'")
+			return fmt.Errorf("cannot find the new commit hash. %s", newOSInfo.CommitID)
+		}
+
+		err = updateGreenbootScripts()
+		if err != nil {
+			log.Error("Failed to update greenboot scripts")
+			return err
+		}
+
+		cmd = exec.Command("rpm-ostree", "upgrade")
+		stdout, err = cmd.Output()
+
+		if err != nil {
+			log.Error("Failed to run 'rpm-ostree upgrade'")
+			return err
+		}
+
+		o.GracefulRebootFlow()
+
+		cmd = exec.Command("systemctl", "reboot")
+		stdout, err = cmd.Output()
+
+		if err != nil {
+			return fmt.Errorf("failed to run 'rpm-ostree reboot': %s", err)
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (o *OS) GracefulRebootFlow(){
+	// send signal for graceful rebooting
+	go func() { o.GracefulRebootChannel <- StartGracefulReboot }()
+	// wait for signal that the graceful reboot FinishedGracefulReboot
+	for {
+		select {
+		case job := <- o.WaitTillFinishGracefulRebootChannel:
+			if job == FinishedGracefulReboot {
+				log.Info("graceful reboot finished")
+				return
+			}
+		case <-time.After(TimeoutToGracefulRebootInSeconds * time.Second):
+			log.Println("Timeout")
+			return
+		}
+	}
 }
 
-func (o *OS) GetOsImageId() string {
-	return UnknownOsImageId
+func (o *OS) GetUpgradeStatus() *models.UpgradeStatus{
+	var upgradeStatus models.UpgradeStatus
+	o.UpdateRpmOstreeStatus()
+	upgradeStatus.CurrentCommitID = o.OsCommit
+	upgradeStatus.LastUpgradeTime = o.LastUpgradeTime
+	upgradeStatus.LastUpgradeStatus = o.LastUpgradeStatus
+
+	return &upgradeStatus
+}
+
+func updateURLInEdgeConfFile(newURL string) error{
+	input, err := ioutil.ReadFile(EdgeConfFileName)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(string(input), "\n")
+
+	for i, line := range lines {
+		if strings.Contains(line, "url=") {
+			lines[i] = "url="+newURL
+		}
+	}
+
+	output := strings.Join(lines, "\n")
+	err = ioutil.WriteFile(EdgeConfFileName, []byte(output), 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updateGreenbootScripts() error{
+	log.Info("Update greenboot scripts")
+	_, err := os.Stat(GreenbootHealthCheckFileName)
+	if err == nil {
+		log.Infof("File %s already exists", GreenbootHealthCheckFileName)
+	} else{
+		if errors.Is(err, os.ErrNotExist) {
+			greenbootHealthCheckFile, err := os.Create(GreenbootHealthCheckFileName)
+			if err != nil {
+				return err
+			}
+			_, err = greenbootHealthCheckFile.Write([]byte(GreenbootHealthCheckScript))
+			if err != nil {
+				return err
+			}
+			greenbootHealthCheckFile.Chmod(755)
+		} else {
+			return err
+		}
+	}
```

Comment:
seems this can be generified and extracted into a method, e.g.: `ensureScriptExists(name, content) err` and reused for both files.